ruby/tests/basic.rb

```diff
@@ -383,10 +383,30 @@ def test_rptfield_array_ducktyping
       length_methods.each do |lm|
         assert l.send(lm)  == 0
       end
+      # out of bounds returns a nil
+      assert l[0] == nil
+      assert l[1] == nil
+      assert l[-1] == nil
+
       l.push 4
       length_methods.each do |lm|
-        assert l.send(lm)  == 1
+        assert l.send(lm) == 1
       end
+      assert l[0] == 4
+      assert l[1] == nil
+      assert l[-1] == 4
+      assert l[-2] == nil
+
+      l.push 2
+      length_methods.each do |lm|
+        assert l.send(lm) == 2
+      end
+      assert l[0] == 4
+      assert l[1] == 2
+      assert l[2] == nil
+      assert l[-1] == 2
+      assert l[-2] == 4
+      assert l[-3] == nil
     end
 
     def test_map_basic
```

Comment:
the test file is getting long; I can keep going in this style, but I can also start pulling them out into separate test files... your call (if I pull out into separate test files, that should probably be its own PR)